zcash_client_backend/src/sync.rs

```diff
@@ -0,0 +1,399 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::anyhow;
+use futures_util::TryStreamExt;
+use prost::Message;
+use tokio::{fs::File, io::AsyncWriteExt, task::JoinHandle};
+use tonic::{
+    body::BoxBody,
+    client::GrpcService,
+    codegen::{Body, Bytes, StdError},
+};
+use tracing::{debug, error, info};
+// TODO: Move `FsBlockDb` behind a `BlockCache` trait to break cyclic dependency.
+use zcash_client_sqlite::{chain::BlockMeta, FsBlockDb, FsBlockDbError};
+use zcash_primitives::{
+    consensus::{BlockHeight, Parameters},
+    merkle_tree::HashSer,
+};
+
+use crate::{
+    data_api::{
+        chain::{error::Error as ChainError, scan_cached_blocks, BlockSource, CommitmentTreeRoot},
+        scanning::{ScanPriority, ScanRange},
+        WalletCommitmentTrees, WalletRead, WalletWrite,
+    },
+    proto::service::{self, compact_tx_streamer_client::CompactTxStreamerClient},
+};
+
+pub(crate) fn get_block_path(fsblockdb_root: &Path, meta: &BlockMeta) -> PathBuf {
+    meta.block_file_path(&fsblockdb_root.join("blocks"))
+}
+
+/// Scans the chain until the wallet is up-to-date.
```

Comment:
Why stop there? Why not add a parameter to indicate that the caller wishes the method to run indefinitely, keeping up with the blockchain?

Reply:
Other parameter ideas from my own implementation: 
- take a callback parameter that receives % progress updates as well as newly found transactions.
- take a cancellation token so the caller can gracefully abort the sync where it's at (such that it can be resumed at roughly the same location later).

Reply:
As I have said in multiple places, this code is extracted from the existing private code that I have (and that I shared previously with you, on top of which you built your implementation). It is intended as a starting point for further development in-repo, not the perfect sync engine up-front.

So the reason why it stops there is that it was simpler for me to implement initially, and also how my CLI wallet works (which has no long-running backend process). Please move these suggestions into separate issues for future discussion and implementation.